PiBridgeMaster.c

```diff
@@ -830,17 +830,25 @@ int PiBridgeMaster_Run(void)
 	}
 	piCore_g.image.drv.i8uStatus = RevPiDevice_getStatus();
 
-	revpi_led_trigger_event(last_led, piCore_g.image.usr.i8uLED);
+	if (piDev_g.machine_type == REVPI_CONNECT_4) {
+		revpi_rgb_led_trigger_event(last_led, piCore_g.image.usr.rgb_leds);
+	} else {
+		revpi_led_trigger_event(last_led, piCore_g.image.usr.leds);
+	}
 	if (piDev_g.machine_type == REVPI_CONNECT ||
 	    piDev_g.machine_type == REVPI_CONNECT_SE) {
-		if ((last_led ^ piCore_g.image.usr.i8uLED) & PICONTROL_X2_DOUT) {
-			gpiod_set_value(piCore_g.gpio_x2do, (piCore_g.image.usr.i8uLED & PICONTROL_X2_DOUT) ? 1 : 0);
+		if ((last_led ^ piCore_g.image.usr.leds) & PICONTROL_X2_DOUT) {
+			gpiod_set_value(piCore_g.gpio_x2do, (piCore_g.image.usr.leds & PICONTROL_X2_DOUT) ? 1 : 0);
 		}
-		if ((last_led ^ piCore_g.image.usr.i8uLED) & PICONTROL_WD_TRIGGER) {
-			gpiod_set_value(piCore_g.gpio_wdtrigger, (piCore_g.image.usr.i8uLED & PICONTROL_WD_TRIGGER) ? 1 : 0);
+		if ((last_led ^ piCore_g.image.usr.leds) & PICONTROL_WD_TRIGGER) {
+			gpiod_set_value(piCore_g.gpio_wdtrigger, (piCore_g.image.usr.leds & PICONTROL_WD_TRIGGER) ? 1 : 0);
 		}
 	}
-	last_led = piCore_g.image.usr.i8uLED;
+	if (piDev_g.machine_type == REVPI_CONNECT_4) {
+		last_led = piCore_g.image.usr.rgb_leds;
+	} else {
+		last_led = piCore_g.image.usr.leds;
+	}
 
 	// update every 1 sec
 	if ((kbUT_getCurrentMs() - last_update) > 1000) {
```

Comment:
There's an issue a bit further below in this function in line 883. Unfortunately the idiotic GitHub UI doesn't let me comment on that line directly. I'm talking about:

```
pI2->usr = pI1->usr;
```

This copies the `usr` portion of the struct. You've just changed its size to 7. So you're copying 7 bytes. But on pre-Connect4 products, the size you want to copy is 5. Hence I think this line needs to be changed to something like:

```
memcpy(&pI2->usr, &pI1->usr, RevPiDevice_getDev(0)->sId.i16uFBS_OutputLength);
```

Which obviously isn't as readable unfortunately.

In practice, the impact of this issue is limited. If there's some DIO directly adjacent to the Core's portion of the process image, you'll copy two bytes from the DIO's process image portion. No biggie, you're not using that extraneous data.

The only case I can imagine when this will become catastrophic is if the Core's portion of the process image is located at the end of the process image. In that case, you'll copy 2 bytes beyond the process image, so an out-of-bounds read, which can lead to a kernel crash. Can that happen right now? It can't because piControl seems to put the Core always at the beginning of the process image. However, if some genius programmer comes along in a few years and moves the Core's portion to the end of the process image, they may end up with a difficult to debug crash.

Thus, for full correctness, something like the above memcpy instruction would be necessary.

@linosanfilippo-kunbus What do you think?

Reply:
Implemented as discussed, compiled and successfully tested on Connect 4 and Core S.